torch/csrc/jit/codegen/cuda/glfdc/eval.cpp

```diff
@@ -0,0 +1,336 @@
+#include <torch/csrc/jit/codegen/cuda/glfdc/eval.h>
+
+#include <torch/csrc/jit/codegen/cuda/glfdc/cfold.h>
+
+using namespace torch::jit::fuser::cuda::glfdc;
+
+namespace {
+
+using skiptree_t = std::vector<std::pair<SExprRef, std::size_t>>;
+
+// Recursive binary tree DFS traversal.
+// Its also counting nodes and values of each subtree;
+template <
+    typename Tree_,
+    typename FnPre_,
+    typename FnPost_,
+    typename LRGetters_>
+size_t dfs_fn(
+    Tree_ t,
+    skiptree_t& list,
+    FnPre_& visit_pre,
+    FnPost_& visit_post,
+    LRGetters_& getters) {
```

Comment:
Do these parameter types really need to be templatized? Is it really necessary to templatize the logic to access tree nodes? It seems to me that the interface is overly generalized, making a simple function non-straightforward to follow. Unless it really needs to be, I'd prefer to make it more trivial and minimally generic.

Reply:
Don't need to be.